api/agent/protocol/json.go

```diff
@@ -0,0 +1,93 @@
+package protocol
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"net/http"
+)
+
+// This is sent into the function
+// All HTTP request headers should be set in env
+type jsonio struct {
+	Headers    http.Header `json:"headers,omitempty"`
+	Body       string      `json:"body"`
+	StatusCode int         `json:"status_code,omitempty"`
+}
+
+// JSONProtocol converts stdin/stdout streams from HTTP into JSON format.
+type JSONProtocol struct {
+	in  io.Writer
+	out io.Reader
+}
+
+func (p *JSONProtocol) IsStreamable() bool {
+	return true
+}
+
+func writeString(err error, dst io.Writer, str string) error {
+	if err != nil {
+		return err
+	}
+	_, err = io.WriteString(dst, str)
+	return err
+}
+
+func (h *JSONProtocol) DumpJSON(req *http.Request) error {
+	stdin := json.NewEncoder(h.in)
+	bb := new(bytes.Buffer)
+	_, err := bb.ReadFrom(req.Body)
+	if err != nil {
+		return err
+	}
+	err = writeString(err, h.in, "{")
+	err = writeString(err, h.in, `"body":`)
+	err = stdin.Encode(bb.String())
+	if err != nil {
```

Comment:
this needs to be in front of the `stdin.Encode` call in order to avoid overwriting any prior errors, not after it